### ISSUE TITLE
Incompatibility with Kotlin-1.7.0-Beta

### ISSUE THREAD

**Issue by WToon** [kotest/kotest#2960]
Using the latest version of Kotest (5.3.0) and the Kotlin 1.7.0-Beta, I am encountering the following error:

```
java.lang.IncompatibleClassChangeError: Found interface kotlin.time.TimeMark, but class was expected
	at io.kotest.engine.spec.interceptor.SpecFinishedInterceptor.intercept-0E7RQCE(SpecFinishedInterceptor.kt:37)
	at io.kotest.engine.spec.interceptor.SpecFinishedInterceptor$intercept$1.invokeSuspend(SpecFinishedInterceptor.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
```

The full stacktrace is attached.

[stacktrace.txt](https://github.com/kotest/kotest/files/8641823/stacktrace.txt)

**Comment by sksamuel** [comment]
Sounds like they've changed the time stuff again.

**Comment by sksamuel** [comment]
We'll prob need to wait for 1.7 final and then cut a 5.4 release that targets that.

**Comment by jandryml** [comment]
I had the same issue with Kotest (5.3.0), but with Kotlin version 1.6.21 also I was using kotlinx-datetime on version 0.3.3. Downgrade of kotlinx-datetime to 0.3.2 solved it for me.

**Comment by WToon** [comment]
> I had the same issue with Kotest (5.3.0), but with Kotlin version 1.6.21 also I was using kotlinx-datetime on version 0.3.3. Downgrade of kotlinx-datetime to 0.3.2 solved it for me.

I tried the following to force version 0.3.2:
`
    implementation("org.jetbrains.kotlinx:kotlinx-datetime:0.3.2"){
        force = true
    }
`
but I get the same error.

**Comment by sksamuel** [comment]
Once 1.7 final is out we will release. 
See https://github.com/kotest/kotest/issues/2990

**Comment by jesperancinha** [comment]
Hi, I just came across this same issue @sksamuel . Can you give an estimate as to when this issue might be solved? Thanks!

**Comment by sksamuel** [comment]
We'll do a 5.4 build once 1.7 is fully out.

**Comment by ismarslomic** [comment]
I'm also facing this issue with kotest `v5.3.0` and kotlin `v1.7.0` (which is available as we speak)

**Comment by sksamuel** [comment]
1.7 just got released a few hours ago but we'll wait for the official blog
post.
Then we can upgrade kotest and make sure it compiles and then release 5.4

**Comment by sschuberth** [comment]
The blog post is there now: https://blog.jetbrains.com/kotlin/2022/06/kotlin-1-7-0-released/

**Comment by aSemy** [comment]
Here's the description of the change: https://kotlinlang.org/docs/whatsnew17.html#time-marks-based-on-inline-classes-for-default-time-source 

(sharing for completeness)

> ### Time marks based on inline classes for default time source﻿
> 
> Kotlin 1.7.0 improves the performance of time measurement functionality by changing the time marks returned by `TimeSource.Monotonic` into inline value classes. This means that calling functions like `markNow()`, `elapsedNow()`, `measureTime()`, and `measureTimedValue()` doesn't allocate wrapper classes for their `TimeMark` instances. Especially when measuring a piece of code that is part of a hot path, this can help minimize the performance impact of the measurement:
> 
> ```kotlin
> @OptIn(ExperimentalTime::class)
> fun main() {
>     val mark = TimeSource.Monotonic.markNow() // Returned `TimeMark` is inline class
>     val elapsedDuration = mark.elapsedNow()
> }
> ```
> 
> This optimization is only available if the time source from which the `TimeMark` is obtained is statically known to be `TimeSource.Monotonic`.

**Comment by sksamuel** [comment]
@aSemy so it sounds like if we don't reference Monotonic, then it will continue to work like before and we don't need to release separately for 1.6 and 1.7

**Comment by sksamuel** [comment]
Doesn't look like other time sources are available :(

**Comment by aSemy** [comment]
Possibly Kotest can try avoiding the call to `measureTimedValue { }`

https://github.com/kotest/kotest/blob/652c768b931b07c95ddf412fee3acbfad224ff66/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/interceptor/SpecFinishedInterceptor.kt#L19

That's the one that references `TimeSource.Monotonic`

https://github.com/JetBrains/kotlin/blob/bcd8a28d4c77d1652ddcaaf0fdbe3524df2a2f05/libraries/stdlib/src/kotlin/time/measureTime.kt#L83

which in turn references the cause: `ValueTimeMark` is now inline

https://github.com/JetBrains/kotlin/blob/bcd8a28d4c77d1652ddcaaf0fdbe3524df2a2f05/libraries/stdlib/src/kotlin/time/TimeSource.kt#L54

EDIT here's an example (I haven't run it)

```kotlin
package io.kotest.engine.spec.interceptor

import io.kotest.core.spec.SpecRef
import io.kotest.core.test.TestCase
import io.kotest.core.test.TestResult
import io.kotest.engine.listener.TestEngineListener
import io.kotest.mpp.timeInMillis
import kotlin.time.Duration.Companion.milliseconds

/**
 * A [SpecRefInterceptor] that invokes the [specFinished] test engine listener callbacks.
 * Any unhandled exception in the spec executor will be handled by this interceptor.
 */
internal class SpecFinishedInterceptor(private val listener: TestEngineListener) : SpecRefInterceptor {

   override suspend fun intercept(
      ref: SpecRef,
      fn: suspend (SpecRef) -> Result<Map<TestCase, TestResult>>
   ): Result<Map<TestCase, TestResult>> {
      val start = timeInMillis()
      val value = fn(ref)
      val duration = (timeInMillis() - start).milliseconds
      return value
         .onSuccess { listener.specFinished(ref.kclass, TestResult.Success(duration)) }
         .onFailure { listener.specFinished(ref.kclass, TestResult.Error(duration, it)) }
   }
}
```

**Comment by ghost** [comment]
Are snapshots working with 1.7?

**Comment by sksamuel** [comment]
5.3.1 has been released, can you try that ?

**Comment by sschuberth** [comment]
Local tests look promising! Let's see what our CI says: https://github.com/oss-review-toolkit/ort/pull/5419

**Comment by jesperancinha** [comment]
@sksamuel , I see that my builds are slowly, but surely, automatically merging again. Looks good! tks! 👍

**Comment by sksamuel** [comment]
Awesome @sschuberth @jesperancinha

**Comment by snowe2010** [comment]
hm.. I'm seeing this error with that release. 

```
java.lang.AbstractMethodError: Receiver class io.kotest.assertions.ErrorCollectorContextElement does not define or inherit an implementation of the resolved method 'abstract kotlinx.coroutines.CopyableThreadContextElement copyForChildCoroutine()' of interface kotlinx.coroutines.CopyableThreadContextElement.

Caused by: java.lang.AbstractMethodError: Receiver class io.kotest.assertions.ErrorCollectorContextElement does not define or inherit an implementation of the resolved method 'abstract kotlinx.coroutines.CopyableThreadContextElement copyForChildCoroutine()' of interface kotlinx.coroutines.CopyableThreadContextElement.
```

**Comment by sksamuel** [comment]
What version of Kotlin and coroutines are you using ?

**Comment by snowe2010** [comment]
```xml
...
        <kotlin.version>1.7.0</kotlin.version>
        <kotest.version>5.3.1</kotest.version>
...
        <dependency>
            <groupId>org.jetbrains.kotlinx</groupId>
            <artifactId>kotlinx-coroutines-core</artifactId>
            <version>1.6.0</version>
        </dependency>
```

Looks like coroutines has a 1.6.2 version, I will try that. I posted that previous message while in the middle of a meeting (posting this one in a meeting too 😅)

**Comment by sksamuel** [comment]
Should work with 1.6.1.

**Comment by nielsvanvelzen** [comment]
There is no published 5.3.1 version of the multiplatform Gradle plugin. Is the version supposed to be in sync with the library?

**Comment by snowe2010** [comment]
had to switch from using `kotlinx-coroutines-core` to `kotlinx-coroutines-core-jvm` but it did work after updating that and the version to `1.6.2`

**Comment by Kantis** [comment]
@nielsvanvelzen it should be out now :)

**Comment by jmfayard** [comment]
I can confirmed this is fixed after updating to Kotlin 1.7.0 and kotest 5.3.1

**Comment by mnrasul** [comment]
@jmfayard Thank you. I can confirm the combo of kotlin `1.7.0` and kotest `5.3.1` works.
